### ISSUE TITLE
ReadiumCSS issue with "cover"  images (resizing in scroll and paginated mode), or integration problem in navigator?

### ISSUE THREAD

**Issue by danielweck** [edrlab/r2-navigator-js#18]
In some cases, there are extra pixels above and below the image (not sure why). Apparently no problems in mobile implementations (iOS-WebKit, Android-Chromium/Blink). To be investigated...

**Comment by JayPanoz** [comment]
By extra pixels, do you mean negative space? Something similar to 

![capture d ecran 2018-12-04 a 12 14 14](https://user-images.githubusercontent.com/12599652/49438284-37fd6380-f7be-11e8-844b-39dfae54b3d5.png)

If so, a possible explanation is the use of `object-fit: contain` to keep the image’s ratio. See MDN demos: https://developer.mozilla.org/en-US/docs/Web/CSS/object-fit

Considering covers are very often sized based on their `height`, it kinda prevents the image from being deformed so if say the `max-width` is reached, then the `height` will be adjusted and you end with this negative space above and below. 

It’s working as `background-size` actually.

**Comment by danielweck** [comment]
Having "blank" spacing above/below or left/right (depending on the image aspect ratio, and assuming the "contain" fit rule) is absolutely fine.
This is a different problem: there is a small amount of vertical scroll in the body (unnecessary top/bottom blank spacing), after the image is "resized" by ReadiumCSS. I have to inspect the DOM etc. to figure out where this is coming from exactly.

**Comment by JayPanoz** [comment]
Thanks for clarifying. Incidentally I spotted this issue as well in webpub-viewer recently but couldn’t give it a proper look. 

On the top of my mind, here are a few clues I wanted to investigate:

1. `img` is impacted by line-height so may have a few pixels in addition to its own sizing;
2. rounding error (but webpub-viewer is appending an height with JS)
3. box-sizing → especially true for `vh` units.

**Comment by llemeurfr** [comment]
It's much more that a few pixels on Readium Desktop ..
![rdsk issue 18](https://user-images.githubusercontent.com/14943186/49444349-2b353b80-f7cf-11e8-880d-0293bc74b499.png)

**Comment by llemeurfr** [comment]
It's much more that a few pixels on Readium Desktop ...

See the right elevator. But the cover appears full.

<img width="799" alt="rdsk issue 18 2" src="https://user-images.githubusercontent.com/14943186/49444492-9f6fdf00-f7cf-11e8-878d-a21f1dc36995.png">

**Comment by JayPanoz** [comment]
Oh yeah that’s definitely too much of an overflow for the first 2 options. Looks like box-sizing. 

So question: 

<img width="799" alt="49444492-9f6fdf00-f7cf-11e8-878d-a21f1dc36995" src="https://user-images.githubusercontent.com/12599652/49449846-127f5280-f7dc-11e8-9eaf-569ce17001a4.png">

Is there any `margin`/`padding` (see blue parts) impacting contents? (`body`, `img`, `figure` or `div`, etc.)

If not, how does the web view interacts with the 2 parts in red?

**Comment by danielweck** [comment]
The navigator viewport is the rectangular area sandwiched between the red parts (thank you for the annotated visuals, by the way).

Therefore, I think that the top red area is part of the readium-desktop UI/chrome. I am not sure whether it is intentional, but it is in fact a `margin-top` of `96px` (`6rem`), also equal to the bottom margin (red area).

CSS culprit:
```css
.content_root {
  height: 100%;
  display: flex;
  align-items: stretch;
  margin-top: 6rem;
  margin-bottom: 6rem;
}
```
https://github.com/readium/readium-desktop/blob/3efc805d2d3458827d50cf341d402c3fda39d0d4/src/renderer/assets/styles/reader-app.css#L72

Applied an ancestor container `div` (`content_root`):
```html
                        <div className={styles.content_root}>
                            <div className={styles.reader}>
                                <div className={styles.publication_viewport_container}>
                                    <div id="publication_viewport" className={styles.publication_viewport}> </div>
                                </div>
                            </div>
</div>
```
https://github.com/readium/readium-desktop/blob/3efc805d2d3458827d50cf341d402c3fda39d0d4/src/renderer/components/reader/ReaderApp.tsx#L330

Let's file a separate issue for this.
https://github.com/readium/readium-desktop/issues/

CC @clebeaupin can you please look into this?

**Comment by danielweck** [comment]
Secondly, the top and bottom blank area above and below the image are the `body`'s own margins (`30px`, as reported by the Chromium web inspector). There is also a padding left/right of `30px`.

Looking at the cascade, this seems to be due to ReadiumCSS's `--RS__pageGutter`. Is there something I should do from the navigator to instruct ReadiumCSS *not* to apply this extra space? Or is that the expected behaviour?

I have not debugged / inspected the Android and iOS apps so I do not know if the generated CSS is the same, and whether or not it affects the user experience (the scroll bar in the desktop app make this pretty obvious).

**Comment by JayPanoz** [comment]
Hmmmmm let me check the commit history in ReadiumCSS. I'm pretty sure that was not originally there and was added after some testing in the iOS app

**Comment by JayPanoz** [comment]
So yeah this is a legacy of the iOS implem:

- original scroll styles are here: https://github.com/readium/readium-css/commit/fd21e53afdad78e750117bd78a2350152c84492f
- vertical margins were added here: https://github.com/readium/readium-css/commit/127b1d3e43813b0b74e4f229bc7d43555aa943af

It’s very likely we added that as the documents’ text would be very near the top and bottom margins (like 4–5 pixels from the edge) on iOS at the time, so it was meant as some little extra room for content to breathe. 

So yeah there are vertical margins there: 

https://github.com/readium/readium-css/blob/051e3d2d18423c1433964a02816f5d93364ec124/css/src/modules/ReadiumCSS-scroll.css#L24-L30

Which can be replaced with `0` – can be the value of the `--RS__pageGutter` variable if you don’t want `padding-left/right` either.

**Comment by danielweck** [comment]
Jiminy, thank you for checking.

So I am still wondering why the iOS and Android apps do not exhibit the same sub-optimal / non-desirable user experience as the desktop app (i.e. unnecessary vertical scroll). Maybe visually it looks fine, but in fact behind the scenes the viewport has some scroll offset?

Side note about ReadiumCSS version: the `package-lock.json` of `readium-desktop` in the `feature/new-db` branch (i.e. current development trunk) references a slightly older revision of `r2-navigator-js`, which has since incorporated the latest code from ReadiumCSS, as detailed here:
https://github.com/readium/r2-navigator-js/tree/develop/misc/ReadiumCSS
https://github.com/readium/r2-navigator-js/blob/develop/CHANGELOG.md#100-alpha7
As usual I will update `package.json` and `package-lock.json` in `readium-desktop` (I do this regularly to make sure "deep" dependencies are also in sync). I actually did propose this upgrade a while ago ( https://github.com/readium/readium-desktop/pull/176 ) but I decided to close the PR due to conflicting `package-lock.json` changes in other branches (`feature/new-ui` and `feature/new-db`). PS: the lightweight `r2-testapp-js` always references the latest `r2-xxx-js` packages, as it is used as the immediate test-bed for any change in underlying components.

**Comment by JayPanoz** [comment]
I just checked iOS (I’ll need to reinstall Android though, as it’s crashing when triggering the user settings) and yeah it overflows as well, only you don’t have a visible scrollbar: 

![photo 04-12-2018 17 23 02](https://user-images.githubusercontent.com/12599652/49456628-aefc2180-f7e9-11e8-9570-c67de88622cf.png)

![photo 04-12-2018 17 23 10](https://user-images.githubusercontent.com/12599652/49456630-aefc2180-f7e9-11e8-99ab-b3e58d4516fa.png)

**Comment by danielweck** [comment]
Thanks a lot Jiminy, much appreciated. I'll have to build a webview-debuggable variant of the iOS and Android apps at my end, so I can inspect DOM + CSS (to compare with the desktop navigator).

**Comment by llemeurfr** [comment]
@danielweck or team up with Aferdita on this mobile side. Can you check with her please?

**Comment by danielweck** [comment]
CC @aferditamuriqi

**Comment by JayPanoz** [comment]
That said, I’m not particularly happy about this, as it was designed to solve an issue with text but doesn’t work very well with images and may even be confusing, as this issue is kinda proving. 

So that’s probably something to discuss collectively, and check what our options are – @ the CSS level, please bear in mind that I must have hooks since I can’t “bubble up” the cascade.

**Comment by JayPanoz** [comment]
BTW, just as a heads-up, they were removed for vertical writing. 

https://github.com/readium/readium-css/blob/051e3d2d18423c1433964a02816f5d93364ec124/css/src/modules/ReadiumCSS-scroll-vertical.css#L24-L26

(In case you’re wondering why `40.01em`, that’s to force a reflow so that scroll mode in `vertical-rl` doesn’t inherit the computed fragmentation from CSS multicol, I’m not kidding.)

**Comment by danielweck** [comment]
Now, I was testing with "Childrens Litterature", which only has an image object in its "cover page":
https://github.com/IDPF/epub3-samples/blob/master/30/childrens-literature/EPUB/cover.xhtml#L9

Same with "Accessible EPUB3", but note the `figure` wrapper, to which ReadiumCSS adds `margin: 1em 5%`:
https://github.com/IDPF/epub3-samples/blob/master/30/accessible_epub_3/EPUB/cover.xhtml#L13
(so, the cumulated effect may result in additional / undesirable scroll extent)

Furthermore, I believe that the "Alice In Wonderland" example contains a heading above the image. I can't find a direct GitHub link, so here is "cc-shared-culture" instead :)
https://github.com/IDPF/epub3-samples/blob/master/30/cc-shared-culture/EPUB/xhtml/cover.xhtml#L14
Naturally, in this case the overflow / scroll extent is greater, and the webview creates an appropriate scroll bar.

**Comment by JayPanoz** [comment]
So yeah this is the complication there: you have different contexts. 

I can recall a discussion with Hadrien and Amy at the very beginning of the project and one question was: how do I remove the margin of the cover if it’s wrapped in a `<p>`? 

That’s not something I can target with CSS (hint: https://caniuse.com/#feat=css-has) and she had done that with JS earlier anyways.

As your testing demonstrates, that can also happen because of HTML5 suggested rendering (we have `margin: 1em 5%` in ReadiumCSS but it’s a slight modification of the default `margin: 1em 40px` that ships with the rendering engine).

So basically, if I don’t have something like `<html style="--image-only: true">` I can’t adapt margins, etc.

**Comment by llemeurfr** [comment]
Is there at least a best practice for publishers? This "cover in content" seems to be a must in many EPUBs.

**Comment by danielweck** [comment]
That's the mystical "cover page" problem (in contrast with "cover image") :)

**Comment by danielweck** [comment]
So, this has clearly turned into a ReadiumCSS conversation. I propose we close this `r2-navigator-js` issue, and copy/paste the relevant / important bits into https://github.com/readium/readium-css/issues/55

**Comment by danielweck** [comment]
I filed a `readium-desktop` issue here:
https://github.com/readium/readium-desktop/issues/185

**Comment by JayPanoz** [comment]
> Is there at least a best practice for publishers? This "cover in content" seems to be a must in many EPUBs.

I mean, they can use `epub:type="cover"` but if it’s on `figure` and not `body`, I can’t remove those margins in CSS because “it keeps going down, it never lifts back to the surface.”

**Comment by danielweck** [comment]
Could you achieve this with the relational pseudo-class `:has()` of CSS Selectors 4?
https://www.w3.org/TR/selectors-4/#relational

**Comment by JayPanoz** [comment]
TL;DR: no, cf. 
https://caniuse.com/#feat=css-has

Long version short: it’s even worse as it is one of those non-critical features that would be W3C-compliant making it to JavaScript – with no matching CSS implementation – and it’s not even implemented there.

**Comment by danielweck** [comment]
Updated ReadiumCSS (no more top/bottom margins):
https://github.com/readium/readium-css/issues/55#issuecomment-481717278
